src/main/java/org/sonarlint/intellij/analysis/LocalFileExclusions.java

```diff
@@ -0,0 +1,215 @@
+        loadGlobalExclusions(newSettings);
+      }
+    });
+    busConnection.subscribe(ProjectConfigurationListener.TOPIC, this::loadProjectExclusions);
+  }
+
+  public Result checkExclusionAutomaticAnalysis(VirtualFile file, @Nullable Module module) {
+    if (!canAnalyze(file, module)) {
+      return Result.excluded(null);
+    }
+
+    if (powerSaveModeCheck.get()) {
+      return Result.excluded("power save mode is enabled");
+    }
+
+    Result r = checkFileInSourceFolders(file, module);
+    if (r.isExcluded) {
+      return r;
+    }
+
+    String relativePath = SonarLintUtils.getRelativePath(module.getProject(), file);
+    if (globalExclusions.test(relativePath)) {
+      return Result.excluded("file matches exclusions defined in the SonarLint Global Settings");
+    }
+    if (projectExclusions.test(relativePath)) {
+      return Result.excluded("file matches exclusions defined in the SonarLint Project Settings");
+    }
+
+    return Result.notExcluded();
+  }
+
+  private Result checkFileInSourceFolders(VirtualFile file, Module module) {
+    ModuleRootManager moduleRootManager = ModuleRootManager.getInstance(module);
+    ContentEntry[] entries = moduleRootManager.getContentEntries();
+
+    for (ContentEntry e : entries) {
+      if (isExcludedOrUnderExcludedDirectory(file, e)) {
+        return Result.excluded("file is excluded in Project Structure");
+      }
+
+      SourceFolder[] sourceFolders = e.getSourceFolders();
+
+      for (SourceFolder sourceFolder : sourceFolders) {
+        if (sourceFolder.getFile() == null || sourceFolder.isSynthetic()) {
+          continue;
+        }
+
+        if (VfsUtil.isAncestor(sourceFolder.getFile(), file, false)) {
+          return checkSourceFolder(sourceFolder);
+        }
+      }
+    }
+
+    // java must be in a source root. For other files, we always analyze them.
+    if ("java".equalsIgnoreCase(file.getFileType().getDefaultExtension())) {
+      return Result.excluded("java file not under a source root");
+    }
+    return Result.notExcluded();
+  }
+
+  private Result checkSourceFolder(SourceFolder sourceFolder) {
+    if (isJavaResource(sourceFolder)) {
+      Result.excluded("file is under a resource folder");
+    } else if (isJavaGeneratedSource(sourceFolder)) {
+      Result.excluded("file belongs to generated source folder");
+    }
+
+    return Result.notExcluded();
+  }
+
+  public boolean canAnalyze(VirtualFile file, @Nullable Module module) {
+    if (module == null) {
+      return false;
+    }
+
+    if (module.isDisposed() || module.getProject().isDisposed()) {
+      return false;
+    }
+
+    if (!file.isInLocalFileSystem() || file.getFileType().isBinary() || !file.isValid()
+      || ".idea".equals(file.getParent().getName())) {
+      return false;
+    }
+
+    // In PHPStorm the same PHP file is analyzed twice (once as PHP file and once as HTML file)
+    if ("html".equalsIgnoreCase(file.getFileType().getName())) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Replace this if-then-else statement by a single return statement. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1126)